lsコマンドを作る5 の課題提出

88-99 opened the pull request:
lsコマンドを作る5 の課題を実装しました。
ご確認お願いいたします。

Commit:
lsコマンドを作る5 のfirst commit

summertree128 commented on 05.ls/lsk:
```diff
@@ -9,33 +9,46 @@ NUMBER_OF_COLUMNS = 3
+end
+
+def search_file_paths(argv_option)
+  {
+    { a: true, l: true, r: true } => make_file_paths(search_dotmatch_filenames.reverse),
+    { a: true, l: true } => make_file_paths(search_dotmatch_filenames),
+    { r: true, l: true } => make_file_paths(Dir.glob('*').reverse),
+    { l: true } => make_file_paths(Dir.glob('*'))
+  }[argv_option]
```
なかなか面白い実装だとは思うのですが、ハッシュのキーにハッシュを使うのはやめといた方が無難かもしれませんね。
参考：https://docs.ruby-lang.org/ja/latest/class/Hash.html#:~:text=%E7%A0%B4%E5%A3%8A%E7%9A%84%E6%93%8D%E4%BD%9C%E3%81%AB%E3%82%88%E3%81%A3%E3%81%A6%E3%82%AD%E3%83%BC%E3%81%A8%E3%81%97%E3%81%A6%E4%B8%8E%E3%81%88%E3%81%9F%E3%82%AA%E3%83%96%E3%82%B8%E3%82%A7%E3%82%AF%E3%83%88%E3%81%AE%E5%86%85%E5%AE%B9%E3%81%8C%E5%A4%89%E5%8C%96%E3%81%97%E3%80%81Object%23hash%20%E3%81%AE%E8%BF%94%E3%81%99%E5%80%A4%E3%81%8C%E5%A4%89%E3%82%8F%E3%82%8B%E3%81%A8%E3%83%8F%E3%83%83%E3%82%B7%E3%83%A5%E3%81%8B%E3%82%89%E5%80%A4%E3%81%8C%E5%8F%96%E3%82%8A%E5%87%BA%E3%81%9B%E3%81%AA%E3%81%8F%E3%81%AA%E3%82%8A%E3%81%BE%E3%81%99%E3%81%8B%E3%82%89%E3%80%81%20Array%E3%80%81Hash%20%E3%81%AA%E3%81%A9%E3%81%AE%E3%82%A4%E3%83%B3%E3%82%B9%E3%82%BF%E3%83%B3%E3%82%B9%E3%81%AF%E3%82%AD%E3%83%BC%E3%81%AB%E5%90%91%E3%81%8D%E3%81%BE%E3%81%9B%E3%82%93%E3%80%82Hash%23rehash%20%E3%82%92%E5%8F%82%E7%85%A7

88-99 replied:
ご指摘ありがとうございます。
case文にして、配列を使ってみましたがいかがでしょうか？

summertree128 commented on 05.ls/lsk:
```diff
@@ -9,33 +9,46 @@ NUMBER_OF_COLUMNS = 3
+end
+
+def search_dotmatch_filenames
+  Dir.glob('*', File::FNM_DOTMATCH)
```
dotmatchじゃない方はメソッド化してないですし、どちらかに揃えた方がいいかなと思いました。個人的には`search_dotmatch_filenames`はメソッド化するほどではないかなーと思います。

88-99 replied:
ありがとうございます。メソッド化をやめました。

summertree128 reviewed (commented):
レビューしました！少しだけコメントしてあるのでご確認ください〜

Commit:
オプションの条件分岐を修正

ハッシュのキーにハッシュはよくないため。

Commit:
コメントアウトの削除

不要なため。

summertree128 commented on 05.ls/lsk:
```diff
@@ -9,33 +9,50 @@ NUMBER_OF_COLUMNS = 3
 
-def search_filenames(_argv_option)
-  Dir.glob('*')
+def search_filenames(argv_option)
+  case argv_option.keys.sort
+  when %i[a r]
+    Dir.glob('*', File::FNM_DOTMATCH).reverse
+  when %i[a]
+    Dir.glob('*', File::FNM_DOTMATCH)
+  when %i[r]
+    Dir.glob('*').reverse
+  when []
+    Dir.glob('*')
+  end
+end
+
+def search_file_paths(argv_option)
+  case argv_option.keys.sort
+  when %i[a l r]
+    make_file_paths(Dir.glob('*', File::FNM_DOTMATCH).reverse)
+  when %i[a l]
+    make_file_paths(Dir.glob('*', File::FNM_DOTMATCH))
+  when %i[l r]
+    make_file_paths(Dir.glob('*').reverse)
+  when %i[l]
+    make_file_paths(Dir.glob('*'))
+  end
```
この二つのメソッドはやっていることがほとんど同じでDRYでないですね。`make_file_paths`をかませるかどうかというところ以外全て同じです。`make_file_paths`を実行するのはこのメソッドの外でやるようにすれば、`search_file_paths`はいらなくなると思います。（あるいは`search_file_paths`の中で`search_filenames`を呼び出せば良い）
DRYなコードにしましょう〜

88-99 replied:
make_file_pathsの後ろ側が、search_filenames(argv_option)と全く同じですね。すみません😅

summertree128 commented on 05.ls/lsk:
```diff
@@ -9,33 +9,50 @@ NUMBER_OF_COLUMNS = 3
+    Dir.glob('*').reverse
+  when []
+    Dir.glob('*')
+  end
```
このような記述をすると、オプションがもう一つ増えた時に`when`の数がかなり多くなってしまいますね。また、あと二つ増えたら？三つ増えたら…？と考えると、いわゆる組合せ爆発が起こってしまいそうです。
組合せ爆発が起こらない書き方にするにはどうすればいいか、考えて修正してみてください〜

88-99 replied:
修正いたしました。
いろいろ試してみましたができなかったので、質問して解決できました。
各オプションを適用するとどうなるのか、ファイル数は変わるのか変わらないのか、それぞれの結果からどの順番で実装していくのが良いか考え、 条件分岐の結果を次の条件分岐に利用していく方法を学びました。

Commit:
オプションの条件分岐と処理方法を変更

変更前は組み合わせをすべてを書き出していたので、オプションが増えるとそれだけ組み合わせが増えてしまう。オプションが増えても少ない範囲で対応できるようにするため。

summertree128 commented on 05.ls/lsk:
```diff
@@ -8,36 +8,28 @@ NUMBER_OF_COLUMNS = 3
   argv_option = parse_option(ARGV)
-  filenames = search_filenames(argv_option)
+  filenames = argv_option[:a] ? Dir.glob('*', File::FNM_DOTMATCH) : Dir.glob('*')
+  filenames = argv_option[:r] ? filenames.reverse : filenames
```
細かいようですが、-r オプションが指定されていない場合に`filenames = filenames`という意味のない処理が走ってしまうのが少し気持ち悪いので、`filenames = filenames.reverse if argv_option[:r]` としたほうがベターかなと思いました。参考まで。

88-99 replied:
ご指摘ありがとうございます🙏
修正してpushしました。
1〜5までのレビューありがとうございました！😊
他の方のコード見て、勉強します。

summertree128 reviewed (approved):
いいと思います！OKです〜

Commit:
rオプションを適用する場合の条件分岐の書き方を変更

rオプションを適用しない場合、filenames = filenamesは意味がないため。
